acme/dns_challenge_cloudflare.go

```diff
@@ -54,85 +66,127 @@ func (c *DNSProviderCloudFlare) Present(domain, token, keyAuth string) error {
 	if err != nil {
-		return records, fmt.Errorf("CloudFlare API call has failed: %v", err)
+		return "", err
 	}
 
-	name := unFqdn(fqdn)
-	for _, rec := range result {
-		if rec.Name == name && rec.Type == "TXT" {
-			records = append(records, rec)
-		}
-	}
-
-	return records, nil
-}
-
-func (c *DNSProviderCloudFlare) getHostedZoneID(fqdn string) (string, error) {
-	zones, err := c.client.Zones.List(c.ctx)
+	var zones []HostedZone
+	err = json.Unmarshal(result, &zones)
 	if err != nil {
-		return "", fmt.Errorf("CloudFlare API call failed: %v", err)
+		return "", err
 	}
 
-	var hostedZone cloudflare.Zone
+	var hostedZone HostedZone
 	for _, zone := range zones {
 		name := toFqdn(zone.Name)
 		if strings.HasSuffix(fqdn, name) {
 			if len(zone.Name) > len(hostedZone.Name) {
-				hostedZone = *zone
+				hostedZone = zone
 			}
 		}
 	}
 	if hostedZone.ID == "" {
-		return "", fmt.Errorf("No matching CloudFlare zone found for domain %s", fqdn)
+		return "", fmt.Errorf("No matching CloudFlare zone found for %s", fqdn)
 	}
 
 	return hostedZone.ID, nil
 }
 
-func newTxtRecord(zoneID, fqdn, value string, ttl int) *cloudflare.Record {
-	name := unFqdn(fqdn)
-	return &cloudflare.Record{
-		Type:    "TXT",
-		Name:    name,
-		Content: value,
-		TTL:     sanitizeTTL(ttl),
-		ZoneID:  zoneID,
+func (c *DNSProviderCloudFlare) findTxtRecord(fqdn string) (*cloudFlareRecord, error) {
+	zoneID, err := c.getHostedZoneID(fqdn)
+	if err != nil {
+		return nil, err
 	}
+
+	result, err := c.makeRequest("GET", fmt.Sprintf("/zones/%s/dns_records?per_page=1000", zoneID), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var records []cloudFlareRecord
+	err = json.Unmarshal(result, &records)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, rec := range records {
+		if rec.Name == unFqdn(fqdn) && rec.Type == "TXT" {
+			return &rec, nil
+		}
+	}
+
+	return nil, fmt.Errorf("No existing record found for %s", fqdn)
 }
 
-// TTL must be between 120 and 86400 seconds
-func sanitizeTTL(ttl int) int {
-	switch {
-	case ttl < 120:
-		return 120
-	case ttl > 86400:
-		return 86400
-	default:
-		return ttl
+func (c *DNSProviderCloudFlare) makeRequest(method, uri string, body io.Reader) (json.RawMessage, error) {
+	// ApiError contains error details for failed requests
+	type ApiError struct {
+		Code    int    `json:"code,omitempty"`
+		Message string `json:"message,omitempty"`
+	}
+
+	// ApiResponse represents a response from CloudFlare API
+	type ApiResponse struct {
+		Success bool            `json:"success"`
+		Errors  []*ApiError     `json:"errors"`
+		Result  json.RawMessage `json:"result"`
+	}
+
+	req, err := http.NewRequest(method, fmt.Sprintf("%s%s", CloudFlareApiURL, uri), body)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("X-Auth-Email", c.authEmail)
+	req.Header.Set("X-Auth-Key", c.authKey)
+	req.Header.Set("User-Agent", userAgent())
+
+	client := http.DefaultClient
+	client.Timeout = time.Duration(30 * time.Second)
```

Comment:
It is dangerous because that variable is shared by all code that uses
net/http Post and Get functions and anyone else using http.DefaultClient,
but the value is not protected by a mutex.

On Thu, Feb 18, 2016, 12:46 PM Jan B <notifications@github.com> wrote:

> In acme/dns_challenge_cloudflare.go
> <https://github.com/xenolf/lego/pull/132#discussion_r53371527>:
>
> > +		Success bool            `json:"success"`
> > +		Errors  []*ApiError     `json:"errors"`
> > +		Result  json.RawMessage `json:"result"`
> > +	}
> > +
> > +	req, err := http.NewRequest(method, fmt.Sprintf("%s%s", CloudFlareApiURL, uri), body)
> > +	if err != nil {
> > +		return nil, err
> > +	}
> > +
> > +	req.Header.Set("X-Auth-Email", c.authEmail)
> > +	req.Header.Set("X-Auth-Key", c.authKey)
> > +	req.Header.Set("User-Agent", userAgent())
> > +
> > +	client := http.DefaultClient
> > +	client.Timeout = time.Duration(30 * time.Second)
>
> DefaultClient is actually just a shortcut for &http.Client{}. So it's not
> dangerous but probably a bit odd to customize the default client. Changed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xenolf/lego/pull/132/files#r53371527>.
>